### ISSUE TITLE
Music randomly stops. (Playtest 20170722)

### ISSUE THREAD

**Issue by AoAGeneral** [OpenRA/OpenRA#13763]
At random points in the game the music will stop as if you pressed paused. IE: Mechanical Man suddenly stops at 0:32 timeframe until you reset the track.

It seems to happen randomly like it used to do with the track skip issue.

**Comment by pchote** [comment]
Ping @RoosterDragon.

**Comment by GraionDilach** [comment]
Thjis happens when too many sound effects are played at once, from what I've noticed.

**Comment by RoosterDragon** [comment]
- When the music pauses on it's own, does pausing it and resuming it via the music menu cause it to pick up where it left off?
- Does it remain paused forever? E.g. If you wait for, let's say, 30 seconds, does it ever resume again on its own?
- Does your game crash, or keep going?

I'm at a loss to explain this one at the moment. In theory the following should all be true:
- If buffering the audio is slow (e.g. your hard drive suddenly becomes super busy or something) then there will be a pause whilst the buffering catches up, but it auto-resumes playing the audio once more is buffered.
- If the audio streaming crashes, then the game as a whole should crash. So if your game doesn't crash, this should indicate the streaming system is still alive and well.
- Regarding too many sound effects, this shouldn't cause music to get stuck part way through a track. If too many sounds are occurring, new sounds are prevented from starting, but existing ones are not stopped.

**Comment by AoAGeneral** [comment]
#1 pausing and resuming does nothing. It has to be stopped and played again.

#2 It remains paused forever. Essentialy it gets stuck.

#3 Game does not crash.

As for buffering and slow hard drives via it being busy the playtest client is on my SSD drive.

**Comment by pchote** [comment]
@RoosterDragon: do you want to put together a branch based off the playtest that writes whatever info would be useful for debugging this to a music.log?  @AoAGeneral could then run it like normal and report back when it happens again.

**Comment by RoosterDragon** [comment]
I have pushed a branch: https://github.com/RoosterDragon/OpenRA/tree/sound-logging. It will create a music.log file in the Logs folder.

When the music stops, it would be good to know the time according to your system clock so I know where I need to dig in the logs.

If it takes a while and the log is large, you can provide just the most recent 5 minutes worth of logging (there are UTC timestamps in the logs that should help) - just be sure not to trim out the log from around the trim the music stopped (it would be good to have, say, at least 30 secs previous of logs to the stoppage). :)

Ping me if you need any help.

**Comment by AoAGeneral** [comment]
Ive done tests on both single player and multiplayer. All be it it wasn't against a player but AI. I did several battles in single player and one that included 31 minutes. I setup a multiplayer game with AI only and did several 1v1 games lasting about 9-15 minutes.

No music skip happened. At this point im going to assume its only on my end this is happening. If this is happening on my end im not sure what the cause is. (Only other option I could come up with is the hardware doesn't like the software but again that would be on my end only.)

**Comment by reaperrr** [comment]
> If too many sounds are occurring, new sounds are prevented from starting, but existing ones are not stopped.

@RoosterDragon 
Out of curiosity, where and how is the limit for 'too many sounds occuring' determined?
Could it be that the streaming mechanism considers the next chunk of the music stream as 'new sound'?

**Comment by pchote** [comment]
Or alternatively/relatedly: could the other sounds be stealing the dequeued buffers out from under the streaming source?

If it is either of these then it would be straightforward to work around by keeping our own count of currently active / total available sources instead of relying on the AL library to do the right thing.

**Comment by RoosterDragon** [comment]
The limit is determined by `PoolSize` in `OpenAlSoundEngine`. When a new sound starts, if the pool has nothing available (after reclaiming any complete sources) then nothing plays. It something is available, it gets to hold onto the source until that sound completes.

When music starts - it owns the source until the music reaches the end, it doesn't have to give back and acquire a new one each time. Each sound (normal sounds and the streaming music) generate their own buffers when they start and then reuse them throughout. The buffers are not shared between sounds.

If music manages to start playing it should in theory have everything it needs until the track finishes, the only things I can currently think of that would stop it are:
- Slow streaming (merely slow streaming should mean the music resumes eventually though, which doesn't seem to be the case here).
- `AL_OUT_OF_MEMORY` when buffering new data to OpenAL (which we don't currently check for). This also seems unlikely though, as if we're low on memory I would fully expect an out-of-memory crash from OpenGL or .Net soon afterwards bringing down the whole program.

**Comment by AoAGeneral** [comment]
Bit curious question, does sound cards have a different operation then compared to onboard sound?

https://www.amazon.com/Blaster-Performance-Headphone-Forming-Microphone/dp/B009ISU33E

Example is this is the sound card I use on my system.

**Comment by GraionDilach** [comment]
I haven't had time so far to do checks with the modified executable - but I can confirm this bug does happen with an onboard audio (I'm not certain about the kind though, the HP 8200 CMT has some sort of a made-for-it mainboard I can't really find specs of).

**Comment by GraionDilach** [comment]
```
2017-08-29T02:04:31.7015946Z Data buffered to OpenAL and queued.
2017-08-29T02:04:31.7015946Z Waiting for a free buffer.
2017-08-29T02:04:32.7019514Z Dequeuing 2 buffers.
2017-08-29T02:04:32.7019514Z Dequeued 2 buffers.
2017-08-29T02:04:32.7019514Z Filling byte buffer from disk.
2017-08-29T02:04:32.7019514Z Filled byte buffer from disk with 44100 bytes. Stream end: False
2017-08-29T02:04:32.7019514Z Data buffered to OpenAL and queued.
2017-08-29T02:04:32.7019514Z Filling byte buffer from disk.
2017-08-29T02:04:32.7019514Z Filled byte buffer from disk with 44100 bytes. Stream end: False
2017-08-29T02:04:32.7019514Z Data buffered to OpenAL and queued.
2017-08-29T02:04:32.7019514Z Waiting for a free buffer.
2017-08-29T02:04:33.7023180Z Dequeuing 0 buffers.
2017-08-29T02:04:33.7023180Z Dequeued 0 buffers.
2017-08-29T02:04:33.7023180Z Waiting for a free buffer.
2017-08-29T02:04:34.7027789Z Dequeuing 0 buffers.
2017-08-29T02:04:34.7027789Z Dequeued 0 buffers.
2017-08-29T02:04:34.7027789Z Waiting for a free buffer.
2017-08-29T02:04:35.7033574Z Dequeuing 0 buffers.
2017-08-29T02:04:35.7033574Z Dequeued 0 buffers.
2017-08-29T02:04:35.7033574Z Waiting for a free buffer.
2017-08-29T02:04:36.7103436Z Dequeuing 0 buffers.
2017-08-29T02:04:36.7103436Z Dequeued 0 buffers.
2017-08-29T02:04:36.7103436Z Waiting for a free buffer.
2017-08-29T02:04:37.7108493Z Dequeuing 0 buffers.
2017-08-29T02:04:37.7108493Z Dequeued 0 buffers.
2017-08-29T02:04:37.7108493Z Waiting for a free buffer.
2017-08-29T02:04:38.7112601Z Dequeuing 0 buffers.
2017-08-29T02:04:38.7112601Z Dequeued 0 buffers.
2017-08-29T02:04:38.7112601Z Waiting for a free buffer.
2017-08-29T02:04:39.7119276Z Dequeuing 0 buffers.
2017-08-29T02:04:39.7119276Z Dequeued 0 buffers.
2017-08-29T02:04:39.7119276Z Waiting for a free buffer.
2017-08-29T02:04:40.7126118Z Dequeuing 0 buffers.
2017-08-29T02:04:40.7126118Z Dequeued 0 buffers.
2017-08-29T02:04:40.7126118Z Waiting for a free buffer.
2017-08-29T02:04:41.7132851Z Dequeuing 0 buffers.
2017-08-29T02:04:41.7132851Z Dequeued 0 buffers.
```

Last three lines loop continously while no music.

**Comment by RoosterDragon** [comment]
My current bet is that one of calls to OpenAL failed for some reason and it's in a bad way. I'll have to add error checks.

Hopefully whatever is sporadically failing can be retried or worked around without just having to kill the track.

**Comment by RoosterDragon** [comment]
I have pushed an update to the branch with comprehensive error checks. Would be grateful if you could run it again and try to reproduce the issue. This will crash if an error is detected so please include the exception logs and well as the music log.

I'm of the opinion we should revert the music streaming if we can't fix it for the release. Music randomly dying is a regression compared to merely a freeze.

**Comment by pchote** [comment]
> I'm of the opinion we should revert the music streaming if we can't fix it for the release. Music randomly dying is a regression compared to merely a freeze.

If it comes to this I'd prefer it if we could add a settings flag instead of disabling it for everyone.  Especially if we can detect the failure case and toggle it automatically.

**Comment by pchote** [comment]
@GraionDilach: I don't recognise the track names in your log, so I assume that they are custom tracks.  If they are in wav format then we expect it to fail because wav does not support streaming - that is a known and separate issue to what is being debugged here.

**Comment by GraionDilach** [comment]
I don't see how the fact they are in wav (they indeed are wav though) cause a different issue when the symptoms are the same and the effect appeared only after the stream PR got merged.

**Comment by pchote** [comment]
Because #13521 explained up front that it didn't support wav, and nobody pressed the point because it didn't affect any shipping mods.

**Comment by GraionDilach** [comment]
The same log is triggered with AUDs as well though - haven't had time to update the debugging build yet - so your claim of WAVs causing it is wrong. It became harder to reproduce with AUDs though, that's all.

**Comment by pchote** [comment]
Ok, that's great then.  Please post your logs ASAP so that we can try and get this fixed before the final release.

**Comment by GraionDilach** [comment]
@RoosterDragon The OpenAL error check didn't caught this issue. The game didn't crashed when the music hit the buffer issue. Basically my first log fragment still applies.

**Comment by pchote** [comment]
Can you please add the following to `DequeueBuffers` (in OpenAlSoundEngine.cs) and log the returned status as part of the "Dequeuing XX buffers" line?

```
int state;
AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
```

**Comment by GraionDilach** [comment]
Will do (the day after) tomorrow. I need more sleep than what I had for the next 24 hours.

**Comment by pchote** [comment]
It would also be interesting to log `SeekPosition` from somewhere in the loop too.

**Comment by RoosterDragon** [comment]
Have pushed a commit with those changes.

Regarding wavs, though they are not streamed from disk, they are still streamed by the audio system. Knowing this affects wavs is helpful - we know it can't be some weird disk read thing - the bytes are coming straight from a `MemoryStream`.

**Comment by GraionDilach** [comment]
Noticed frozen track, skipped to next and prev imminently to force-replay to see if it gets stuck again.

https://gist.github.com/GraionDilach/6a354771bb34df4c53a2a8e7da2e0257

**Comment by pchote** [comment]
For what it's worth, this is what I see on macOS:

https://gist.github.com/pchote/a97174786c9ecf562091b2a04cd4ee97

This is the same issue that I reported and investigated in #13521, except that it happens all the time with wavs, without modifying delays.  It may or may not be related to the problem here.

**Comment by RoosterDragon** [comment]
Graion's issue appears to be that stops, and swallows the buffers (if the track stops, then all buffers should be reported as processed). Since it just reports no buffers, they're effectively lost. We can't even safely free them and queue up more as you have to dequeue them before you can free them.

pchote, your issue appears to be the same as before, it just keeps "plays" straight through the buffer without actually playing the audio.

At this point I'm thinking would should drop the streaming sound source - it seems like the implementations aren't too reliable.

**Comment by pchote** [comment]
If we do drop it, then we should at least keep the music loading on a background thread to avoid the large and annoying jank when switching tracks.  We'll need to make a decision on this ASAP, since we want to do another and hopefully final playtest soon.

**Comment by AoAGeneral** [comment]
This game wasn't played on RoosterDragon's feature since I haven't run into the issue seemingly by playing against bots or single player matches. (Although this could be possible.) On the new playtest the music did the freeze at the 7:48 mark. The only clear thing that seemingly happened that I could see was a husk exploding in anjew's base. https://www.dropbox.com/s/reumpo65zv4uci8/OpenRA-2017-09-07T165915Z.orarep?dl=0

**Comment by pchote** [comment]
This may be an openal-soft bug.  The [changelog for 1.18.1](https://github.com/kcat/openal-soft/blob/master/ChangeLog) mentions:
> Fixed an issue where resuming a source might not restart playing it.

We are currently using 1.16.0.

@AoAGeneral: can you please replace `soft_oal.dll` in the OpenRA dir with the newer version from http://kcat.strangesoft.net/openal-binaries/openal-soft-1.18.1-bin.zip  (in the `bin/Win32` dir) and see if it fixes the problem? @GraionDilach: you'll need to compile from source if you're on Linux.

**Comment by VolkovChitzkoi** [comment]
I can state to have found the same problem as AoAGeneral. It also seems that two play-test ago this problem wasn't present.

**Comment by pchote** [comment]
@VolkovChitzkoi: can you please try the potential fix i suggested just above too?

**Comment by VolkovChitzkoi** [comment]
I substitue such file and launcher Tiberian Dawn. It gave me a fatal error:

exception-2017-09-12T170510Z.log:

OpenRA engine version playtest-20170902
Date: 2017-09-12 17:05:10Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.InvalidOperationException`: Failed to initialize platform-integration library. Check graphics.log for details.
   in OpenRA.Game.Initialize(Arguments args)
   in OpenRA.Program.Run(String[] args)
   in OpenRA.Program.Main(String[] args)

graphics.log:

System.BadImageFormatException: Tentativo di caricare un programma con un formato non corretto. (Eccezione da HRESULT: 0x8007000B)
   in OpenAL.ALC10.alcOpenDevice(String devicename)
   in OpenRA.Platforms.Default.OpenAlSoundEngine..ctor(String deviceName)
   in OpenRA.Platforms.Default.DefaultPlatform.CreateSound(String device)
   in OpenRA.Sound..ctor(IPlatform platform, SoundSettings soundSettings)
   in OpenRA.Game.Initialize(Arguments args)

**Comment by reaperrr** [comment]
Fwiw, the dll causes no problems on my system.

**Comment by RoosterDragon** [comment]
Have a stopgap that removes the streaming in #14017, but keeps the other changes. Hopefully we can revisit true streaming at some point once we narrow down if the issue lies with us or with the libraries.

**Comment by GraionDilach** [comment]
@VolkovChitzkoi Have you started it with the x64 dll? OpenRA is compiled as x86 on Windows, so reőplacing the soft_oal.dll with the x64 build within that zip can cause that issue.

I'll be doing stress-tests with that version though.

**Comment by pchote** [comment]
@GraionDilach: @AoAGeneral confirmed on IRC that upgrading `soft_oal.dll` does not help.  #14017 is going to be our best, and probably only, option for the coming playtest and release.

**Comment by GraionDilach** [comment]
Don't want to be the party-pooper but I can again reproduce this in bleed with WAV. However, there's a significant difference now - the game DOES switch to the next track on it's own when the music would've end anyway and doesn't stop completely.

Will probably restore the debug music commits and then open a new bug with the results.

**Comment by RoosterDragon** [comment]
Updated the https://github.com/RoosterDragon/OpenRA/tree/sound-logging branch for you. Let me know if that yields anything interesting.
